src/source_files.jl

```diff
@@ -25,31 +28,32 @@ function SourceFile(code::AbstractString; filename=nothing)
     if isempty(code) || last(code) != '\n'
         push!(line_starts, ncodeunits(code)+1)
     end
-    SourceFile(code, filename, line_starts)
+    SourceFile(code, filename, lineno, line_starts)
 end
 
-function SourceFile(; filename)
-    SourceFile(read(filename, String); filename=filename)
+function SourceFile(; filename, kwargs...)
+    SourceFile(read(filename, String); filename=filename, kwargs...)
 end
 
 # Get line number of the given byte within the code
-function source_line(source::SourceFile, byte_index)
-    line = searchsortedlast(source.line_starts, byte_index)
-    return (line < lastindex(source.line_starts)) ? line : line-1
+function source_line_index(source::SourceFile, byte_index)
+    lineidx = searchsortedlast(source.line_starts, byte_index)
+    return (lineidx < lastindex(source.line_starts)) ? lineidx : lineidx-1
 end
+source_line(source::SourceFile, byte_index, lineidx=source_line_index(source, byte_index)) = lineidx + source.lineno - 1
```

Comment:
We don't want the extra parameter to be part of the public API do we?

```suggestion
_source_line(source::SourceFile, lineidx) = lineidx + source.lineno - 1
source_line(source::SourceFile, byte_index) = _source_line(source, source_line_index(source, byte_index))
```